datastore/datastore/storagebackend/postgresql/putobservations.go

```diff
@@ -75,14 +82,42 @@ func getTSColVals(tsMdata *datastore.TSMetadata) ([]interface{}, error) {
 	return colVals, nil
 }
 
-// getTimeSeriesID retrieves the ID of the row in table time_series that matches tsMdata,
-// inserting a new row if necessary. The ID is first looked up in a cache in order to save
-// unnecessary database access.
+// getTSColValsUnique gets the subset of colVals that correspond to the fields defined by
+// constraint unique_main.
+//
+// The order in the returned array is consistent with the array returned by getTSMdataColsUnique().
+//
+// Returns (column values, nil) upon success, otherwise (..., error).
+func getTSColValsUnique(colVals map[string]interface{}) ([]interface{}, error) {
+
+	result := []interface{}{}
+
+	for _, col := range getTSMdataColsUnique() {
+		colVal, found := colVals[col]
+		if !found {
+			return []interface{}{}, fmt.Errorf("column '%s' not found in colVals: %v", col, colVals)
+		}
+		result = append(result, colVal)
+	}
+
+	return result, nil
+}
+
+// upsertTS retrieves the ID of the row in table time_series that matches tsMdata wrt.
+// the fields - U - defined by constraint unique_main, inserting a new row if necessary.
+//
+// If the row already existed, the function ensures that the row is updated with the tsMdata
+// fields - UC - that are not in U (i.e. the complement of U).
+//
+// The ID is first looked up in a cache (where the key consists of all fields (U + UC)) in order to
+// save unnecessary database access.
+//
 // Returns (ID, nil) upon success, otherwise (..., error).
-func getTimeSeriesID(
+func upsertTS(
 	db *sql.DB, tsMdata *datastore.TSMetadata, cache map[string]int64) (int64, error) {
 
-	colVals, err := getTSColVals(tsMdata)
+	colVals2 := map[string]interface{}{}
+	colVals, err := getTSColVals(tsMdata, colVals2) // column values for U+UC
```

Comment:
Asymmetry in how `colVals` and `colVals2` are returned. Return both as return values?

Or, why not return `colVals` and `colsValsUnique` directly from `getTsColVals()`, both as slice? So that you don't have to call `getTSColValsUnique` later?

Reply:
I agree with the first comment, so will return both as return values. Change coming up.

But I don't see the value of refactoring to derive colVals and colValsUnique from the same function, neither in terms of performance or code readability. (But if they were derived from the same function to begin with, I probably wouldn't have bothered to split into two functions).

Reply:
The reason for the second point is that `getTSColVals` function now returns the column values in two formats, a slice and a dictionary. This feels wrong to me. As the dict is only used to make the colValsUnique, I suggested to return that directly, restoring symmetry. Another option is to let `getTSColCals` return only the dict, and derive the two slices from that.

Reply:
But, it's not a big deal. :-)

Reply:
Well I see your point more clearly now. And I agree. I'll see if I can improve it.

Reply:
I like your suggestion of letting getTSColCals return only the dict, and derive the two slices from that. So that's what I'll implement.

Reply:
But then I ran into an issue probably caused by golang randomizing map iteration. So will have to sort ...

Reply:
I'm afraid I have to revert to the original implementation of having getTSColVals return two structures: one map and one array. The problem is that other parts of the code assumes a predefined order of the ts fields (in particular that the non-string fields (not handleable with reflection) need to be processed explicitly in a certain order (so far the links field is the only one in this category)). Returning just a map and then deriving the two slices from that messes up the order. It's probably possible to fix this somehow, bit I suspect it will take a while to get it right everywhere, and I propose to do this in a later refactoring. So being a bit pragmatic for now. I will add a comment about it in the code, though.

Reply:
Hm ... looking again at your earlier suggestion:

> Or, why not return colVals and colsValsUnique directly from getTsColVals(), both as slice? So that you don't have to call getTSColValsUnique later?

I'll try that. Seems to make good sense.

Reply:
(Sorry for once again getting your point quite late in the conversation :) )